File: Python/Unique Binary Search Trees II.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @return a list of tree node
    def generateTrees(self, n):
        return self.generate(1,n)
        
    
    def generate(self, left , right):
        if left == right: return[TreeNode(left)]
        if left > right: return [None]
        
        ret = []
        for i in range(left, right+1):
            lefttree = self.generate(left,i-1)
            righttree = self.generate(i+1, right)
            for l in lefttree:
                for r in righttree:
                    node = TreeNode(i)
                    node.left = l
                    node.right = r
                    ret.append(node)
        return ret
        
        

File: Java/src/Unique Binary Search Trees II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; left = null; right = null; }
 * }
 */
public class Solution {
    public List<TreeNode> generateTrees(int n) {
        return helper(1,n);
    }
    public List<TreeNode> helper(int left, int right){
        List<TreeNode> ret = new ArrayList<>();
        
        if (left == right){
            ret.add(new TreeNode(left));
            return ret;
        } 
        if (left > right) {
            ret.add(null);
            return ret;
        }
        
        for (int i = left; i<= right; i++){
            List<TreeNode> lefttree = helper(left, i-1);
            List<TreeNode> righttree = helper(i+1, right);
            
            for(TreeNode l: lefttree){
                for(TreeNode r: righttree){
                    TreeNode node = new TreeNode(i);
                    node.left = l;
                    node.right = r;
                    ret.add(node);
                }
            }
        }
        return ret;
    }
}